server/bin/test_data_importer.py

```diff
@@ -0,0 +1,865 @@
+
+        for owner_key, key_name, pool in activation_keys:
+            log.info('  Activation key: %s [owner: %s, pool: %s]', key_name, owner_key,
+                pool['id'] if pool is not None else 'n/a')
+
+            key = cp.create_activation_key(owner_key, {'name': key_name})
+            if pool is not None:
+                cp.add_pool_to_activation_key(key['id'], pool['id'])
+
+
+# Process a given JSON file, creating the test data it defines
+def process_file(cp, options, filename):
+    with open(filename) as file:
+        data = json.load(file)
+
+    log.info('Importing data from file: %s', filename)
+
+    # Create owners
+    owners = create_owners(cp, data)
+
+    # Create users
+    create_users(cp, data)
+
+    # Create roles
+    create_roles(cp, data)
+
+    # Create products
+    create_products(cp, data)
+
+    # Create pools/subscriptions
+    create_subscriptions(cp, options, owners, data)
+
+    # Create activation keys
+    create_activation_keys(cp, owners)
+
+    # Done!
+    log.info('')
+    log.info('Finished importing data from file: %s', filename)
+
+# Parse CLI options
+def parse_options():
+    parser = argparse.ArgumentParser(description='Imports JSON-formatted test data to a Candlepin server', add_help=True)
+
+    parser.add_argument('--debug', action='store_true', default=False,
+        help='Enables debug output')
+
+    parser.add_argument('--username', action='store', default='admin',
+        help='The username to use when making requests to Candlepin; defaults to \'admin\'')
+    parser.add_argument('--password', action='store', default='admin',
+        help='The password to use when making requests to Candlepin; defaults to \'admin\'')
+    parser.add_argument('--host', action='store', default='localhost',
+        help='The hostname/address of the Candlepin server; defaults to \'localhost\'')
+    parser.add_argument('--port', action='store', default=8443, type=int,
+        help='The port to use when connecting to Candlepin; defaults to 8443')
+    parser.add_argument('--prefix', action='store', default='candlepin',
+        help='The endpoint prefix to use on the destination Candlepin server; defaults to \'candlepin\'')
+
+    parser.add_argument('--cert_dir', action='store', default='generated_certs',
+        help='The directory in which to store generated product certificates; defaults to \'generated_certs\'')
+
+    parser.add_argument("json_files", nargs='*', action='store')
+
+    options = parser.parse_args()
+
+    # Set logging level as appropriate
+    if options.debug:
+        log.setLevel(logging.DEBUG)
+
+    if len(options.json_files) < 1:
+        log.debug("No test data files provided, defaulting to 'test_data.json'")
+
+        filename = 'test_data.json'
+        if sys.path[0]:
+            filename = '{script_home}/{filename}'.format(script_home=sys.path[0], filename=filename)
+
+        options.json_files = [filename]
+
+    return options
+
+
+# Execute script
```

Comment:
I'll add this, but for clarity: I don't intend for this to be imported elsewhere, so that wasn't a concern when this started.

As far as the Candlepin class goes: it exists because we currently have nothing for that at present, and we have other initiatives that should eventually supersede the need to manually write one later (OpenAPI/spec first). The Ruby script got to piggyback on the existing spec tests having a generic-ish candlepin_api class to use for sending everything, but we don't yet have something equivalent in Python. As I don't want to write one here, I implemented the bare minimum needed for the test data creation, but in a way that if/when it gets replaced with a full thing later, we can delete the Candlepin class from this file and import the fully-featured version with minimal pain.

Reply:
Thanks, good to know.